### PR TITLE
Add basic class for simple resource events

### PR DESCRIPTION
**Pull request by initvector**
This update adds a very simple class that will be used for creating [PSR-14-style events](https://www.php-fig.org/psr/psr-14/) for basic resource operations (e.g. adding a discussion, deleting a comment, registering a user).

### CODE REVIEW

File: library/Garden/Events/ResourceEvent.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    public function getPayload(): ?array {
+        return $this->payload;
+    }
+
+    /**
+     * Get the event type.
+     *
+     * @return string|null
+     */
+    public function getType(): ?string {
```
Comment (tburry) [inline library/Garden/Events/ResourceEvent.php]: I think this should just be `string` to avoid edge cases. Also, the constructor makes it a string.

File: library/Garden/Events/ResourceEvent.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        $this->payload = $payload;
+        $this->type = $this->typeFromClass();
+    }
+
+    /**
+     * Get the event action.
+     *
+     * @return string|null
+     */
+    public function getAction(): ?string {
```
Comment (tburry) [inline library/Garden/Events/ResourceEvent.php]: I think this should just be `string` to avoid edge cases. Also, the constructor makes it a string.

**Comment by tburry** [comment]
You can add some basic tests for this class to ensure that the arguments passed to the constructor come back in the getters. This is something @blutarch can do if you wish.

File: library/Garden/Events/ResourceEvent.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+namespace Garden\Events;
+
+/**
+ * An event affecting a specific resource.
+ */
+abstract class ResourceEvent {
+
+    /** A resource has been removed. */
+    const ACTION_DELETE = "delete";
```
Comment (tburry) [inline library/Garden/Events/ResourceEvent.php]: Now that we are PHP 7.1 it is a best practice to declare these as `public`.
